app-ng/src/job.lisp

```diff
@@ -0,0 +1,99 @@
+  (status 'fraiche :type job-status)
+  error)
+
+(defmacro with-job-lock ((job &key (wait t)) &body body)
+  "Evaluate BODY with the LOCK acquired for the duration of BODY. If WAIT is T, block until the lock is available."
+  `(progn
+     (bt:acquire-lock (job-lock ,job) ,wait)
+     ,@body
+     (bt:release-lock (job-lock ,job))))
+
+(defmethod print-object ((job job) stream)
+  (print-unreadable-object (job stream :type nil :identity t)
+    (format stream "QVM Job")))
+
+(defun job-running-p (job)
+  "Is the JOB in the RUNNING state."
+  (with-job-lock (job)
+    (eq (job-status job) 'running)))
+
+(defun job-interrupted-p (job)
+  "Is the JOB in the INTERRUPTED state?"
+  (with-job-lock (job)
+    (eq (job-status job) 'interrupted)))
+
+(defun job-finished-p (job)
+  "Is the JOB in the FINISHED state?"
+  (with-job-lock (job)
+    (eq (job-status job) 'finished)))
+
+(defun make-job (fn)
+  "Make a JOB object which will, under JOB-START, run FN in a thread.
+
+The returned JOB can be repeatedly started if the JOB-STATUS is FINISHED (or INTERRUPTED). Only interact with the job after acquiring the lock (or alternatively by using WITH-JOB-LOCK)."
+  (let* ((job (%make-job))
+         (thr-fn (lambda ()
+                   (setf (job-thread job)
+                         (bt:make-thread
+                          (lambda ()
+                            (handler-case
+                                (progn
+                                  ;; Reset run-specific slots
+                                  (setf (job-%result job) nil)
+                                  (setf (job-error   job) nil)
+                                  (setf (job-status  job) 'running)
+                                  (setf (job-%result job) (funcall fn))
+                                  (setf (job-status  job) 'finished))
+                              (error (c)
+                                ;; TODO What happens to a thread when
+                                ;; there is an error in it?
+                                (setf (job-error job) c)
+                                (setf (job-status job) 'error)))))))))
+    (setf (job-threader job) thr-fn)
+    (setf (job-lock job) (bt:make-lock))
+    job))
+
+;;; CUSTODIAL
+
+(defun job-start (job)
+  "Start the JOB and return T. If JOB is already running, do nothing and return NIL."
+  (with-job-lock (job)
+    (unless (job-running-p job)
+      (funcall (job-threader job))
+      t)))
+
+(defun job-stop (job &key (status 'interrupted))
+  "Forcefully stop JOB if running. Does not block. Set the JOB-STATUS into STATUS."
+  (check-type status job-status)
+  (with-job-lock (job)
+    (when (job-running-p job)
+      (bt:destroy-thread (job-thread job))
+      (setf (job-status job) status))))
+
+;;; sync/async
+
+(defun job-result (job)
+  "Inspect the result of JOB. If JOB is in the running state, this will block."
+  (with-job-lock (job)
+    (when (job-running-p job)
```

Comment:
What if I invert the `with-job-lock` before `job-running-p` structure? 

e.g.
```
(defun job-stop (job)
  "Forcefully stop JOB if running. Does not block."
  (with-job-lock (job)
    (when (job-running-p job)
      (bt:destroy-thread (job-thread job))
      (setf (job-status job) 'interrupted))))
```
becomes
```
(defun job-stop (job)
  "Forcefully stop JOB if running. Does not block."
  (when (job-running-p job)
    (with-job-lock (job)
      (bt:destroy-thread (job-thread job))
      (setf (job-status job) 'interrupted))))
```

Reply:
No bueno, I think. What happens if job status changes between the time when `job-running-p` returns and the following `with-job-lock` executes?

Reply:
@appleby 😿

Reply:
What if these predicates/accessors do not make any claim to be safe? i.e., callers must hold the lock themselves?

Reply:
Concurrency, he's a harsh mistress.

Reply:
Nice catch, @appleby!

I'd favor the use of recursive locks or reading the slot directly. Creating non-(b)locking versions of the functions looks like it could lead to errors down the line.

Reply:
In POSIX threads there are read-write locks that let multiple threads read concurrently but only allow a single thread to hold the lock for writing. That functionality doesn't seem to be available in Bordeaux Threads, though.

Reply:
> What if these predicates/accessors do not make any claim to be safe? i.e., callers must hold the lock themselves?

Also fine, I guess, but in that case I would prefix with `%` or suffix with `-unsafe` or something, to make it obvious these are "internal" functions that expect the caller to handle locking. If possible, it's nice to hide the locking "inside" this file (or package if it were ever to be refactored), so that anything external can treat the locking as an implementation detail it doesn't need to worry about, and only call thread-safe functions this file provides. Even if it's not always possible to avoid requiring the caller to handle locking themselves, it's nice to provide the thread-safe versions in case someone just want to check the status and doesn't need to acquire the lock for anything else.

Reply:
@jmbr I think as long as the thread-safe versions are the "default" with standard naming and the non-locking versions are named some way that makes it obvious to the caller they must first acquire the lock, it's not too bad in practice. Maybe I've just been corrupted by seeing this idiom in lots of C code though!

I also threw out the idea of read/write locks in an offline thread with @notmgsk a while back, then also realized they aren't supported by bordeaux-threads. One of the downsides of using a portable common lisp compatibility layer like `bt` is you wind up with some unsatisfying minimal intersection of supported features between all implementations.